src/NumericalAlgorithms/Spectral/SwshCollocation.hpp

```diff
@@ -0,0 +1,256 @@
+
+ private:
+  // an extra pointer layer is required for the peculiar way that libsharp
+  // constructs these values.
+  std::unique_ptr<sharp_geom_info*, detail::DestroySharpGeometry> geom_info_;
+  size_t l_max_ = 0;
+};
+
+bool operator==(typename Collocation<ComplexRepresentation::RealsThenImags>::
+                    CollocationIterator lhs,
+                typename Collocation<ComplexRepresentation::RealsThenImags>::
+                    CollocationIterator rhs) noexcept;
+bool operator!=(typename Collocation<ComplexRepresentation::RealsThenImags>::
+                    CollocationIterator lhs,
+                typename Collocation<ComplexRepresentation::RealsThenImags>::
+                    CollocationIterator rhs) noexcept;
+
+bool operator==(
+    typename Collocation<
+        ComplexRepresentation::Interleaved>::CollocationIterator lhs,
+    typename Collocation<
+        ComplexRepresentation::Interleaved>::CollocationIterator rhs) noexcept;
+bool operator!=(
+    typename Collocation<
+        ComplexRepresentation::Interleaved>::CollocationIterator lhs,
+    typename Collocation<
+        ComplexRepresentation::Interleaved>::CollocationIterator rhs) noexcept;
+
+namespace detail {
+// This function is threadsafe because of the very limited use of the static
+// variable. It is guaranteed that the initialization of the static variable
+// will be performed exactly once across threads.
+template <ComplexRepresentation Representation, size_t LMax>
+Collocation<Representation>& precomputed_static_collocation_impl() noexcept {
+  static Collocation<Representation> precomputed_collocation =
+      Collocation<Representation>{LMax};
+  return precomputed_collocation;
+}
+
+template <ComplexRepresentation Representation, size_t... Is>
+SPECTRE_ALWAYS_INLINE Collocation<Representation>&
+dispatch_to_precomputed_static_collocation_impl(
+    size_t index, std::index_sequence<Is...> /*meta*/) noexcept {
+  Collocation<Representation>* collocation;
+  auto get_collocation_if_match = [&collocation, &index ](auto i) noexcept {
+    if (UNLIKELY(decltype(i)::value == index)) {
+      collocation =
+          &(precomputed_static_collocation_impl<Representation,
+                                                decltype(i)::value>());
+    }
+    return 0;
+  };
+  expand_pack(
+      get_collocation_if_match(std::integral_constant<size_t, Is>{})...);
+  return *collocation;
+}
+}  // namespace detail
+
+/// \brief precomputation function for those collocation grids that are
+/// requested
+///
+/// \details keeps a static `unordered_map` of all collocation grids that
+/// have been requested, so that the libsharp generation need not be re-run.
+/// If it has been generated, it's returned by reference. Otherwise, the new
+/// grid is generated and put in the dictionary before it is returned by
+/// reference.
+template <ComplexRepresentation Representation>
+Collocation<Representation>& precomputed_spherical_harmonic_collocation(
```

Comment:
This looks bad... You should never return non-const references to cached things because then people can mutate the cache, which seems scary. Also, here it makes the code no longer thread-safe.